### PR TITLE
[`flake8-comprehensions`] autofix C414 and C417 + bugfix

### PR DESCRIPTION
**Pull request by sbrugman**
Closes https://github.com/charliermarsh/ruff/issues/2262 and closes https://github.com/charliermarsh/ruff/issues/2423

Fixes bug where some cases generated duplicated violations (see https://github.com/charliermarsh/ruff/pull/2732#issuecomment-1426397842)

### CODE REVIEW

File: src/rules/flake8_comprehensions/fixes.rs [hunk 1]
```diff
@@ -875,3 +878,198 @@ pub fn fix_unnecessary_comprehension(
...
+    expr: &rustpython_ast::Expr,
+    kind: &str,
+) -> Result<Fix> {
+    let module_text = locator.slice_source_code_range(&Range::from_located(expr));
+    let mut tree = match_module(module_text)?;
+    let mut body = match_expr(&mut tree)?;
+    let call = match_call(body)?;
+    let arg = match_arg(call)?;
+
+    // TODO: Unwrap if list, set or dict comprehension
```
Comment (sbrugman) [inline src/rules/flake8_comprehensions/fixes.rs:934]: Resolved, remove comment

File: src/rules/flake8_comprehensions/rules.rs [hunk 2]
```diff
@@ -678,10 +682,35 @@ pub fn unnecessary_map(checker: &mut Checker, expr: &Expr, func: &Expr, args: &[
...
+                if let ExprKind::Call { func: f, .. } = &parent.node {
+                    let Some(id_parent) = function_name(f)  else {
+                        return;
+                    };
+                    if id_parent == "dict" || id_parent == "set" || id_parent == "list" {
+                        return;
+                    }
+                };
+            };
+            // TODO: exclude if parent is list, set or dict call
```
Comment (sbrugman) [inline src/rules/flake8_comprehensions/rules.rs:696]: Done, remove comment

**Comment by sbrugman** [comment]
Comment by @spaceone 
> C417 doesn't detect all cases anymore. It fixes:
> 
> ```diff
> -        assert all(map(lambda v: isinstance(v, dict), iterator))
> +        assert all((isinstance(v, dict) for v in iterator))
> ```
> 
> better would be:
> 
> ```diff
> -        assert all(map(lambda v: isinstance(v, dict), iterator))
> +        assert all(isinstance(v, dict) for v in iterator)
> ```

**Comment by spaceone** [comment]
with the MR the following is not detected anymore as `C417`:
```python
bitstring = ''.join(map(lambda x: x in logontimes and '1' or '0', range(168)))
```

**Comment by sbrugman** [comment]
Will have a look after the rebase

File: crates/ruff/src/rules/flake8_comprehensions/rules/unnecessary_map.rs [hunk 3]
```diff
@@ -1,18 +1,35 @@
...
+    /// `map(func, iterable)` has great performance when func is a built-in function, and it
+    /// makes sense if your function already has a name. But if your func is a lambda, it’s
+    /// faster to use a generator expression or a comprehension, as it avoids the function call
+    /// overhead. For example:
+    ///
+    /// Rewrite `map(lambda x: x + 1, iterable)` to `(x + 1 for x in iterable)`
+    /// Rewrite `map(lambda item: get_id(item), items)` to `(get_id(item) for item in items)`
+    /// Rewrite `list(map(lambda num: num * 2, nums))` to `[num * 2 for num in nums]`
+    /// Rewrite `set(map(lambda num: num % 2 == 0, nums))` to `{num % 2 == 0 for num in nums}`
+    /// Rewrite `dict(map(lambda v: (v, v ** 2), values))` to `{v : v ** 2 for v in values}`
```
Comment (spaceone) [inline crates/ruff/src/rules/flake8_comprehensions/rules/unnecessary_map.rs:27]:
```suggestion
    /// Rewrite `dict(map(lambda v: (v, v ** 2), values))` to `{v: v ** 2 for v in values}`
```

**Pull request merged**